src/System.Linq.Expressions/src/System/Runtime/CompilerServices/ReadOnlyCollectionBuilder.cs

```diff
@@ -462,15 +526,18 @@ public ReadOnlyCollection<T> ToReadOnlyCollection()
             return new TrueReadOnlyCollection<T>(items);
         }
 
+        // Ensures that the capacity of this list is at least the given minimum
+        // value. If the currect capacity of the list is less than min, the
+        // capacity is increased to twice the current capacity or to min,
+        // whichever is larger.
         private void EnsureCapacity(int min)
         {
             if (_items.Length < min)
             {
-                int newCapacity = DefaultCapacity;
-                if (_items.Length > 0)
-                {
-                    newCapacity = _items.Length * 2;
-                }
+                int newCapacity = _items.Length > 0 ? _items.Length * 2 : DefaultCapacity;
+                // Allow the list to grow to maximum possible capacity (~2G elements) before encountering overflow.
+                // Note that this check works even when _items.Length overflowed thanks to the (uint) cast
+                if ((uint)newCapacity > int.MaxValue) newCapacity = int.MaxValue;
```

Comment:
I know about this magic number 0X7FEFFFFF but it is not public. It will be out off memory exception if request more then  0X7FEFFFFF items.